core/src/mock_alpenglow_consensus.rs

```diff
@@ -0,0 +1,593 @@
+            (
+                "percent_stake_collected_finalize_cert",
+                percent_collected[3],
+                f64
+            ),
+            (
+                "weighted_delay_ms_finalize_cert",
+                stake_weighted_delay[3],
+                f64
+            ),
+        );
+    }
+
+    pub(crate) fn signal_new_slot(&mut self, slot: Slot, interval: Slot) {
+        // constraint interval to be above 3 to keep current logic working
+        let interval = interval.max(3);
+        let phase = slot % interval;
+        // this is currently set up to work over a course of 3 slots.
+        // it will only work correctly if we do not actually fork within those 3 slots, and
+        // are consistently casting votes for all 3.
+        match phase {
+            // Clear the stats for the previous slots in preparation for the new one
+            0 => {
+                // prepare to receive votes for the next slot in advance
+                // in case we are really late getting shreds
+                self.prepare_to_receive(slot + 1);
+            }
+            // send Notarize votes (rest of the consensus happens based on received packets)
+            1 => {
+                trace!("Starting voting in slot {slot}");
+                self.command_sender.send(Command::SendNotarize(slot));
+            }
+            // report metrics
+            2 => {
+                trace!("Reporting statistics for slot {}", slot - 1);
+                // avoid holding locks while reporting metrics
+                let (peers, total_staked) = {
+                    let mut lockguard = self.state.lock().unwrap();
+                    let total_staked = lockguard.total_staked;
+                    let peers = lockguard.reset();
+                    (peers, total_staked)
+                };
+                Self::report_collected_votes(&peers, total_staked);
+            }
+            // in case we have missed previous slot, clean up and block reception
+            // but do not report stats (as they will be all garbled by now)
+            _ => {
+                let mut lockguard = self.state.lock().unwrap();
+                // block reception of votes
+                lockguard.reset();
+            }
+        }
+    }
+
+    pub(crate) fn join(self) -> thread::Result<()> {
+        self.state.lock().unwrap().should_exit = true;
+        self.sender_thread.join()?;
+        self.listener_thread.join()
+    }
+}
+
+fn prep_and_sign_packet(
+    packet_buf: &mut [u8],
+    slot: Slot,
+    id: &Pubkey,
+    state: VotorMessageType,
+    keypair: &Keypair,
+) {
+    // prepare the packet to send and sign it
+    {
+        let pkt = MockVotePacketHeader::from_bytes_mut(packet_buf);
+        pkt.slot_number = slot;
+        pkt.sender = *id.as_array();
+        pkt.signature = [0; SIGNATURE_BYTES];
+        pkt.state = state as u64;
+    }
+    let signature = keypair.sign_message(&packet_buf[SIGNATURE_BYTES..]);
```

Comment:
Do we need this? This uses ed25519 signature anyway so the cost is totally different from BLS. Are we doing this so people can't fake these packets?

Reply:
Yes, it is to prevent people from having fun with these. I've not made up my mind 100% about whether we need it or not. If we have BLS code already merged in agave we can use that and get an idea about realworld perf.

Reply:
We do have bls crate https://github.com/anza-xyz/solana-sdk/blob/master/bls-signatures/Cargo.toml

But BLS verification is much more expensive than ed25519, and we don't have BLS pubkeys for everyone in vote account yet, so let's stick to ed25519 for now.